### DISCUSSION TITLE
Does this library allow for structured MDC data?

### DISCUSSION THREAD

**Discussion by namero999** [rfoltyns/log4j2-elasticsearch#64] [Q&A] [answered]
I couldn't find any reference to MDC in the docs of the HTTP client, and I cannot see the MDC of my threads being logged. So I wonder if I need to configure it somehow or if it's not supported yet. Thanks!

**Comment by rfoltyns** [comment, marked as answer]
Yes, it does. Have a look at [`JacksonJsonLayout` documentation](https://github.com/rfoltyns/log4j2-elasticsearch/tree/master/log4j2-elasticsearch-core#jacksonjsonlayout).

MDC values are resolved with Log4j2 lookups (actually StrSubstitutor under the hood). `VirtualProperty` - optional and configurable within `JacksonJsonLayout` component - wraps StrSubstitutor and supports variable resolution just like Log4j2 KeyPair. It's quite limited though:
* `${ctx:variableName}` can be used only with synchronous loggers
* log event properties e.g. `${event:Marker}` are not supported

If you'd like to resolve MDC values, try this:
```xml
<JacksonJsonLayout>
    <VirtualProperty name="myMdcField" value="$${ctx:propertyName:-undefined}"/>
</JacksonJsonLayout>
```

**Reply by namero999** [reply]
Thanks, I managed to have it work by adding `dynamic="true"` to `VirtualProperty`

**Reply by namero999** [reply]
Is there any specific reason why this is limited to sync loggers?

**Reply by rfoltyns** [reply]
> Is there any specific reason why this is limited to sync loggers?

StrSubstitutor gets the `ctx` values from `ContextDataInjector`. Log4j2 AsyncLogger doesn't copy ThreadContext data to this injector, so values are not available when `VirtualPropertyWriter.valueResolver` tries to resolve it.

**Reply by namero999** [reply]
Interesting, thanks. I've done some research and I've read that starting from recent versions of log4j2 it is recommended to use `ContextDataProvider` instead of `ContextDataInjector` https://logging.apache.org/log4j/2.x/manual/extending.html#Custom_ContextDataProvider

I'd look into it but navigating the codebase of this repo is a bit overwhelming right now. Any pointer?

**Reply by rfoltyns** [reply]
It's out of scope of this repo. This library handles Log4j2 LogEvent(s) on the appender level only.

Setting ThreadContext values in `ContextDataInjector` happens on logger level, before the event reaches any appenders, then it's passed as event.contextData only.

Have a look at variable resolution at [VirtualPropertyWriter:138](https://github.com/rfoltyns/log4j2-elasticsearch/blob/6924f5fd81c8cc23d5b27bfc044221771183c651/log4j2-elasticsearch-core/src/main/java/org/appenders/log4j2/elasticsearch/VirtualPropertiesWriter.java#L138) all the way down to relevant lookup in log4j2-core [ContextMapLookup](https://github.com/apache/logging-log4j2/blob/e917bac53f42191bceaed08e076d5dfe624d1133/log4j-core/src/main/java/org/apache/logging/log4j/core/lookup/ContextMapLookup.java#L43) and [Interpolator](https://github.com/apache/logging-log4j2/blob/e917bac53f42191bceaed08e076d5dfe624d1133/log4j-core/src/main/java/org/apache/logging/log4j/core/lookup/Interpolator.java#L222).

Once you'll dig into data available in AsyncLogger's disruptor consumer thread, you'll notice that event context data is available only at event level, `ContextMapLookup.currentContextData()` is empty and `ValueResolver` (by design) doesn't have access to it.

Fortunately, you can still serialize event.contextData as described [here](https://github.com/rfoltyns/log4j2-elasticsearch/issues/18#issuecomment-488107973). In that case `VirtualProperty` is obsolete.
